### PR TITLE
[WIP] 5 spack script for teuchos

### PR DESCRIPTION
**Pull request by mperrinel**
@marcinwrobel1986 @ppebay @thearusable @MikolajZuzek

First version of teuchos spack file.
1) Permits to run "spack install teuchos"
2) Permits to propagate optional packages to the cmake configuration using variants.

Current list of variant :
 Name [Default]                 When    Allowed values          Description
    ===========================    ====    ====================    ==================================

    ARPREC [off]                   --      on, off                 Compile with ARPREC
    BinUtils [off]                 --      on, off                 Compile with BinUtils
    Boost [off]                    --      on, off                 Compile with Boost
    Eigen [off]                    --      on, off                 Compile with Eigen
    MPI [off]                      --      on, off                 Compile with MPI parallelism
    QD [off]                       --      on, off                 Compile with QD
    Qt [off]                       --      on, off                 Compile with Qt
    build_type [RelWithDebInfo]    --      Debug, Release,         CMake build type
                                           RelWithDebInfo,         
                                           MinSizeRel              
    ipo [off]                      --      on, off                 CMake interprocedural optimization

### CODE REVIEW

**Comment by mzuzek** [comment]
@mperrinel @marcinwrobel1986
Would it be possible to rebase this PR on some `develop` _without_ merging it here ?
(this PR contains now 118 commits that change 244 files...)

**Comment by marcinwrobel1986** [comment]
> @mperrinel @marcinwrobel1986 Would it be possible to rebase this PR on some `develop` _without_ merging it here ? (this PR contains now 118 commits that change 244 files...)

@mzuzek I fetched the upstream and updated the repository. But Spack is updated very often, so it's hard to do that everyday or even a few times a day. This was one of the reason I would like to have that in our repository, not to update this one constantly...

**Comment by mperrinel** [comment]
@mzuzek @marcinwrobel1986, it is not a good process to rebase a PR every day on the main develop stream. We don't really care about days to days spack update. 
1) we should have a stable fork develop branch based on the main spack develop branch. This branch has to be rebased on the main develop branch when we started the project (1 month ago) and when we want to create a PR from our develop to the main develop branch of spack. In addition, if we need a specific new feature of spack that exists since less than a month, we can also exceptionnaly rebase our fork develop branch.
2) I agree with @mzuzek that merging strategy is not good and we should keep the rebase strategy 
3) Our feature branch should be based on our develop branch which can be stable the time of the project, which means that we really don't need to rebase it every day. The branch is created on the top of develop and then rebased on develop (if it has changed) when opening the PR and eventually before merging the PR if our fork develop branch has changed between these two moments)

Eventually, as I proposed and we did for TrilinosEZ, we can create a develop-ez branch to clarify the process and let the develop follows the main develop branch

**Comment by mzuzek** [comment]
@mperrinel @marcinwrobel1986
I'm sorry, `develop` and rebasing was not my point actually, so let me rephrase:
can we please clean up this PR (or replace with new one)
so it does _not_ contain 118 commits changing 244 files and can be reviewed ?

**Comment by marcinwrobel1986** [comment]
Moved to other repository, closing
